flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/ConfigurationNotAllowedError.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.dispatcher;
+
+import org.apache.flink.annotation.Internal;
+
+/**
+ * If {@link org.apache.flink.configuration.DeploymentOptions#ALLOW_CLIENT_JOB_CONFIGURATIONS} is
+ * disabled this error denotes the not allowed configuration.
+ */
+@Internal
+public class ConfigurationNotAllowedError implements JobValidationError {
```

Comment:
I think I can understand why you did it this way, because it feels like it contains some kinds of "business logic" of the user code. Alternatively, have you considered to leverage the Java unchecked exception concept? The Error classes defined here look and act as same as an unchecked exception. User code that is running in Flink could also be considered as a part program logic of Flink and IMHO still a good fit for using Java unchecked exception.

Reply:
I would argue the other way around. Program logic that is buggy has to be expected from Flink's perspective. Executing such logic should not bring Flink to its knees but instead simply complete the job in a `FAILED` state.

Reply:
The other way works too. I think there is no conflict. Flink and user jobs are in different scopes. Consider Flink as the platform and the job with user code is a piece of dynamic logic running on it. The unchecked exception mentioned here will only bring the dynamic logic piece to its knees within the scope of the user code. Flink plays the controller role outside of this scope to manage the exceptions and complete the dynamic logic(the job) and mark it as `FAILED`, in the meantime, Flink itself is still stable and healthy, i.e. any these kinds of unchecked exception will not effect Flink itself.

Reply:
Speaking of "Program logic that is buggy has to be expected from Flink's perspective", if I understood you correctly, agree with you, should use checked exception.

Reply:
One thing I see is one could convert the error classes to real exceptions and use `ExceptionUtils#firstOrSuppressed` but I am not if this makes it harder for a user to digest.